Documentation for new request pipeline life-cycle hooks/p… (#2008)
apollographql/apollo-server: docs/source/features/plugins.md

    },
  };
};
```

Within the `plugins` array, this `localPluginModule.js` would be used as:

```js
  /* ... Existing, required ApolloServer configuration. ... */

  plugins: [

    require('./localPluginModule')({
      /* ...configuration options, when necessary! */
    }),

  ],

  /* ... any additional ApolloServer configuration. ... */
```

And finally, advanced cases can implement the `ApolloServerPlugin` interface via a factory function.  The factory function will receive `pluginInfo`, which can allow implementors to adjust their behavior based on circumstantial factors:

```js
/* advancedPluginModule.js */
module.exports = (options) => {
  /* ...Plugin specific implementation... */

  return () => {

    return {
      requestDidStart() {
        console.log('The options were', options);
      }
    };
  };
}
```

And again, this could be used as a plugin by defining it in the `plugins` array:

```js
  /* ... Existing, required ApolloServer configuration. ... */

  plugins: [

    require('./advancedPluginModule')({
      /* ...configuration options, when necessary! */
    }),

  ],

  /* ... any additional ApolloServer configuration. ... */
```

> > **Note:** Currently the `pluginInfo` is undefined, but future additions to the plugin API will enable this functionality.  For now, the factory function facilities are in place to use, but `pluginInfo` is simply not available.

Author: ```suggestion
```

---
title: Plugins
description: Extending Apollo Server through the use of plugins.
---

> **Note:** Plugins are available in Apollo Server 2.2.x or higher.

## Overview

The default Apollo Server installation is designed for a reliable out-of-the-box experience with as little configuration as possible.  In order to provide additional features, Apollo Server supports _plugins_ which can interface with various stages of server operation and each request/response cycle.

> 

Author: ```suggestion

For example, these life-cycle hooks might be used to:

* Change the HTTP response code depending on the result of the operation's execution using the `willSendResponse` life-cycle hook. 
* Implement rate-limiting for certain IP addresses using the `requestDidStart` life-cycle hook.
```

## Usage

Plugins for Apollo Server can be specified using the `plugins` configuration parameter to the Apollo Server constructor options.

The `plugins` array is an array of plugins.  They might be provided as a module (and optionally published to a registry — e.g. npm) or defined in-line within the `ApolloServer` constructor.  Plugins should be defined correctly and the requirements of building a plugin are explained in the plugin [definition](#definition) section below.

> > **Note:** If a plugin is provided by a package published to a registry (for example, npm), that package must be installed using `npm install <plugin>` or `yarn add <plugin>` prior to use.  In-line plugins or plugins which reside locally do not need to be installed.

Author: ```suggestion
```

```js
module.exports = {
  requestDidStart() {
    /* ... */
  },
};
```

Plugins which accepted options might providing a function which returns an object that implements a object matching the `ApolloServerPlugin` interface:

```js
/* localPluginModule.js */
module.exports = (options) => {
  /* ...Plugin specific implementation... */

  return {
    requestDidStart() {
      console.log('The options were', options);
    },
  };
};
```

Within the `plugins` array, this `localPluginModule.js` would be used as:

```js
  /* ... Existing, required ApolloServer configuration. ... */

  plugins: [

    require('./localPluginModule')({
      /* ...configuration options, when necessary! */
    }),

  ],

  /* ... any additional ApolloServer configuration. ... */
```

And finally, advanced cases can implement the `ApolloServerPlugin` interface via a factory function.  The factory function will receive `pluginInfo`, which can allow implementors to adjust their behavior based on circumstantial factors:

```js
/* advancedPluginModule.js */
module.exports = (options) => {
  /* ...Plugin specific implementation... */

>   return (pluginInfo) => {

Author: ```suggestion
  return () => {
```

>     console.log('The pluginInfo was', pluginInfo);

Author: ```suggestion
```

    return {
      requestDidStart() {
        console.log('The options were', options);
      }
    };
  };
}
```

And again, this could be used as a plugin by defining it in the `plugins` array:

```js
  /* ... Existing, required ApolloServer configuration. ... */

  plugins: [

    require('./advancedPluginModule')({
      /* ...configuration options, when necessary! */
    }),

  ],

  /* ... any additional ApolloServer configuration. ... */
```

> > **Note:** Currently the `pluginInfo` is undefined, but future additions to the plugin API will enable this functionality.  For now, the factory function facilities are in place to use, but `pluginInfo` is simply not available.

Reviewer: Let's comment out the pluginInfo section with a todo until it is available. When do you anticipate we'll be able to use that functionality?

Author: Not sure!  It's reserved for future use, though it probably makes sense to get something — anything — in there sooner rather than later! 😄

Author: Removed for now!

> ## Events

Reviewer: I think this section should be higher up the hierarchy chain, possibly even above the definition, since it's core to understanding plugins. Then, we can pull `Server lifecycle events` and `Request lifecycle events` into their own headings with all of the lifecycle methods as subheadings. Right now, it looks strange in the deploy preview that all of the methods are on the same hierarchy level as `Server lifecycle events` and `Request lifecycle events`.

> There are two main categories of events: server life-cycle events and request life-cycle events.

Reviewer: Do we want to drop the dash in between life-cycle? In the React docs, they spell it without the dash. https://reactjs.org/docs/state-and-lifecycle.html

Author: Let's drop it!  (It's hard to say; `life-cycle` has almost a billion matches on Google and `lifecycle` only has 128 million.  Hard to know if "lifecycle" is more closely associated with CS, but I digress.)

My vote would be that we do it consistently across all our documentation, and based on [this search](https://www.google.com/search?q=site%3Aapollographql.com+life-cycle), that would be `lifecycle`.  I'll make the change in a follow-up commit.

const server = new ApolloServer({
  /* ... other necessary configuration ... */

  plugins: [
    {
      /* The `requestDidStart` will be called when the request has started
         processing and more granular events — like `parsingDidStart` below —
         are executed when those particular events occur. */
      requestDidStart(requestContext) {

        /* Request-specific scope can be created here and
           used in more granular life-cycle events below. */

        return {

          parsingDidStart(requestContext) {
            /* This `parsingDidStart` life-cycle event is
               called when parsing begins, but scoped within the
               `requestDidStart` server life-cycle event. */
          },

        }
      }
    }
  ],

  /* */
})
```

If there are no specific request life-cycle events to implement, `requestDidStart` should not return anything.

### Request life-cycle events

Request life-cycle events must be implemented by returning an object which defines their behavior from the `requestDidStart` server life-cycle event.  By maintaining this structure, coupling logic, and defining plugin-specific request scope becomes semantic and co-located.

> > **Types:** The `apollo-server-plugin-base` module exports [the `GraphQLRequestListener` interface](https://github.com/apollographql/apollo-server/blob/master/packages/apollo-server-plugin-base/src/index.ts) which defines the shape of request life-cycle events.  It's recommended to use this interafce when building custom plugins which implement granular request life-cycle events via `requestDidStart`.

Reviewer: ```suggestion
> **Types:** The `apollo-server-plugin-base` module exports [the `GraphQLRequestListener` interface](https://github.com/apollographql/apollo-server/blob/master/packages/apollo-server-plugin-base/src/index.ts) which defines the shape of request life-cycle events.  It's recommended to use this interface when building custom plugins in TypeScript which implement granular request life-cycle events via `requestDidStart`.
```

#### TypeScript signature

```typescript
validationDidStart?(
  requestContext: WithRequired<GraphQLRequestContext<TContext>, 'document'>,
): (err?: ReadonlyArray<Error>) => void | void;
```

### `didResolveOperation`

The `didResolveOperation` request lifecycle event is triggered after the operation to be executed has been successfully retrieved from the `document` AST by `graphql`'s `getOperationAST`.  This focusing of execution which identifies the correct operation is important when a `document` contains multiple operations.  At this stage, in addition to the `document` the `operationName` (`String`) and `operation` (AST) will be present on the context.

If the operation is anonymous (e.g. the operation is `query { ... }` rather than `query NamedQuery { ... }`, then `operationName` will be `null`.

```typescript
didResolveOperation?(
  requestContext: WithRequired<
    GraphQLRequestContext<TContext>,
    'document' | 'operationName' | 'operation'
  >,
): ValueOrPromise<void>;
```

### `didEncounterErrors`

> TODO

```typescript

> didEncounterErrors?(

Reviewer 2: Can this be used in place of `formatErrors` because it is async and has access to context?

Author: Also, yes!

  requestContext: WithRequired<
    GraphQLRequestContext<TContext>,
    'metrics' | 'source' | 'errors'
  >,
): ValueOrPromise<void>;
```

### `executionDidStart`

> TODO

```typescript
executionDidStart?(
  requestContext: WithRequired<
    GraphQLRequestContext<TContext>,
    'document' | 'operationName' | 'operation'
  >,
): (err?: Error) => void | void;
```

### `willSendResponse`

> TODO

```typescript

> willSendResponse?(

Reviewer 2: can this be used to modify the response before returning, such as adding additional metrics?

Author: Yes!

Reviewer 3: Thanks for including this. If we wanted to alter the extensions object in the response, would either/both of these approaches work?

```
// Approach 1
const helloPlugin1 = {

  requestDidStart() {

    return {
      willSendResponse(requestContext) {

        requestContext.response.extensions.randomField = 'Hello World!'

      },
    };
  },
};

// Approach 2
const _ = require('lodash')

const helloPlugin2 = {

  requestDidStart() {

    return {
      willSendResponse(requestContext) {

        const newReqContext = _.cloneDeep(requestContext);

        newReqContext.response.extensions.randomField = 'Hello World!'

        return newReqContext
      },
    };
  },
};

```
